MSBuild should allow CI tools to isolate invocations

To be robust CI tools need to keep build/test processes from interfering with each other:
* Cleanup left over processes from old runs
* Kill process trees after a timeout

The way I have implemented this is to (pseudo code):
```
// Launch process with CREATE_SUSPENDED
...

// Setup sandbox job object
HANDLE SandboxJob = CreateJobObjectW(nullptr, JobName);
AssignProcessToJobObject(SandboxJob, pi.hProcess);

JOBOBJECT_EXTENDED_LIMIT_INFORMATION LimitInfo;
LimitInfo.BasicLimitInformation.LimitFlags |= JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE;
SetInformationJobObject(Job, JobObjectExtendedLimitInformation, &LimitInfo, sizeof(LimitInfo));

ResumeThread(pi.hThread);
// Let process run until exit
...

// Cleanup any processes left in job
TerminateJobObject(SandboxJob, 255);
```

The problem you run into with MSBuild is that even if you specify `/nodeReuse:false` it will still share nodes between MSBuild.exe invocations.

### Current  behavior
```
Process             Job Object
MSBuild.exe         Build 1
    MSBuild.exe     Build 1
    MSBuild.exe     Build 1
    MSBuild.exe     Build 1
    MSBuild.exe     Build 1
MSBuild.exe         Build 2   < Uses nodes created in Build 1
```
After the first MSBuild.exe exits all nodes will also be killed and the second build will fail with:
`MSBUILD : error MSB4166: Child node "7" exited prematurely`

### Proposed behavior
To workaround this I previously injected code into the build process to make the MSBuild communication pipes unique. This has worked well, but the solution is problematic and prone to break as APIs and software evolve.

A relatively simple fix for this would be to allow the CI tool to make the communication pipes unique. See: https://github.com/Malterlib/msbuild/commit/079b2e1abf5a0fdb058e0d287bf6a86730d49438
```
Process             Job Object
MSBuild.exe         Build 1    < CI tool have set a unique MSBUILD_PIPE_NAME for this job object
    MSBuild.exe     Build 1
    MSBuild.exe     Build 1
    MSBuild.exe     Build 1
    MSBuild.exe     Build 1
MSBuild.exe         Build 2    < CI tool have set a unique MSBUILD_PIPE_NAME for this job object
    MSBuild.exe     Build 2
    MSBuild.exe     Build 2
    MSBuild.exe     Build 2
    MSBuild.exe     Build 2
```
Another option could be to exclude node reuse if you detect that the other node is inside another job object.

This is on version 15.6.82.30579

This is what the build looks like with the unpatched version:
![officialmsbuild](https://user-images.githubusercontent.com/7736494/38076754-db486654-3336-11e8-8412-3a272e7d5c2a.png)

Notice that the lower invocation has 14 child nodes, while the top one has 12 nodes (this is a 12 core machine). 

I have suspended the link process of the one with 12 nodes to force the other one to finish first. After this build finished it promptly failed the other build with:
```
MSBUILD : error MSB4166: Child node "6" exited prematurely. Shutting down. Diagnostic information may be found in files in the temporary files directory named MSBuild_*.failure.txt.
```

This is the same build process with the patched MSBuild (this is on top of the vs15.7 branch):
![patchedmsbuild](https://user-images.githubusercontent.com/7736494/38077020-c47b85ae-3337-11e8-84ff-5acd71f90995.png)

This time both invocations have 13 child nodes each, and after the non-suspended build has finished the other build finished successfully after it has been resumed.

When these builds are started they start pretty much simultaneously, so possibly you have a race condition of some kind in the node reuse logic.

For a fix, consider that it should cope with nodes being terminated at any point, such as during the handshake process, so it might be a good idea to sidestep the issue by changing the pipe name instead of adding a condition in the handshake process.

--- a/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs
+++ b/src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs
@@ -172,33 +172,37 @@ namespace Microsoft.Build.BackEnd
             }
 
 #if FEATURE_NODE_REUSE
-            var candidateProcesses = GetPossibleRunningNodes(msbuildLocation);
-
-            CommunicationsUtilities.Trace("Attempting to connect to each existing msbuild.exe process in turn to establish node {0}...", nodeId);
-            foreach (Process nodeProcess in candidateProcesses)
+            // Try to connect to idle nodes if node reuse is enabled.
+            if (_componentHost.BuildParameters.EnableNodeReuse)
             {
-                if (nodeProcess.Id == Process.GetCurrentProcess().Id)
-                {
-                    continue;
-                }
+                var candidateProcesses = GetPossibleRunningNodes(msbuildLocation);
 
-                // Get the full context of this inspection so that we can always skip this process when we have the same taskhost context
-                string nodeLookupKey = GetProcessesToIgnoreKey(hostHandshake, clientHandshake, nodeProcess.Id);
-                if (_processesToIgnore.Contains(nodeLookupKey))
+                CommunicationsUtilities.Trace("Attempting to connect to each existing msbuild.exe process in turn to establish node {0}...", nodeId);
+                foreach (Process nodeProcess in candidateProcesses)
                 {
-                    continue;
-                }
+                    if (nodeProcess.Id == Process.GetCurrentProcess().Id)
+                    {
+                        continue;
+                    }
 
-                // We don't need to check this again
-                _processesToIgnore.Add(nodeLookupKey);
+                    // Get the full context of this inspection so that we can always skip this process when we have the same taskhost context
+                    string nodeLookupKey = GetProcessesToIgnoreKey(hostHandshake, clientHandshake, nodeProcess.Id);
+                    if (_processesToIgnore.Contains(nodeLookupKey))
+                    {
+                        continue;
+                    }
 
-                // Attempt to connect to each process in turn.
-                Stream nodeStream = TryConnectToProcess(nodeProcess.Id, 0 /* poll, don't wait for connections */, hostHandshake, clientHandshake);
-                if (nodeStream != null)
-                {
-                    // Connection successful, use this node.
-                    CommunicationsUtilities.Trace("Successfully connected to existed node {0} which is PID {1}", nodeId, nodeProcess.Id);
-                    return new NodeContext(nodeId, nodeProcess.Id, nodeStream, factory, terminateNode);
+                    // We don't need to check this again
+                    _processesToIgnore.Add(nodeLookupKey);
+
+                    // Attempt to connect to each process in turn.
+                    Stream nodeStream = TryConnectToProcess(nodeProcess.Id, 0 /* poll, don't wait for connections */, hostHandshake, clientHandshake);
+                    if (nodeStream != null)
+                    {
+                        // Connection successful, use this node.
+                        CommunicationsUtilities.Trace("Successfully connected to existed node {0} which is PID {1}", nodeId, nodeProcess.Id);
+                        return new NodeContext(nodeId, nodeProcess.Id, nodeStream, factory, terminateNode);
+                    }
                 }
             }
 #endif

